File: apps/scala/design-pattern/src/Command.scala
// Command パターン
// trait とケースクラスで操作のオブジェクト化・Undo を実現する

package designpattern.command

import scala.collection.mutable.ListBuffer

trait Command:
  def execute(): Unit
  def undo(): Unit
  def description: String

class SliderDocument:
  private val _content: ListBuffer[String] = ListBuffer.empty

  def content: String = _content.mkString

  def insertString(position: Int, text: String): Unit =
    _content.insert(position, text)

  def deleteString(position: Int, length: Int): Unit =
    for _ <- 0 until length do
      if position < _content.length then _content.remove(position)

class InsertCommand(document: SliderDocument, position: Int, text: String) extends Command:
  override def execute(): Unit =
    document.insertString(position, text)

  override def undo(): Unit =
    document.deleteString(position, text.length)

  override val description: String = s"Insert '$text' at position $position"

class DeleteCommand(document: SliderDocument, position: Int, text: String) extends Command:
  override def execute(): Unit =
    document.deleteString(position, text.length)

  override def undo(): Unit =
    document.insertString(position, text)

  override val description: String = s"Delete '$text' at position $position"

class CompositeCommand(commands: List[Command]) extends Command:
  override def execute(): Unit = commands.foreach(_.execute())
  override def undo(): Unit = commands.reverse.foreach(_.undo())
  override val description: String = commands.map(_.description).mkString("; ")

class CommandHistory:
  private val history: ListBuffer[Command] = ListBuffer.empty

  def executeCommand(command: Command): Unit =
    command.execute()
    history += command

  def undoLast(): Option[Command] =
    if history.nonEmpty then
      val last = history.remove(history.length - 1)
      last.undo()
      Some(last)
    else
      None

  def size: Int = history.size

File: apps/javascript/design-pattern/src/command.js
// Command パターン
// 要求をオブジェクトとしてカプセル化し、実行・取り消しを可能にする

import { writeFileSync, unlinkSync, readFileSync, existsSync } from 'fs';

export class CreateFileCommand {
  constructor(filePath, content) {
    this.filePath = filePath;
    this.content = content;
    this.description = `ファイル作成: ${filePath}`;
  }

  execute() {
    writeFileSync(this.filePath, this.content, 'utf-8');
  }

  undo() {
    if (existsSync(this.filePath)) {
      unlinkSync(this.filePath);
    }
  }
}

export class DeleteFileCommand {
  constructor(filePath) {
    this.filePath = filePath;
    this.description = `ファイル削除: ${filePath}`;
    this._backup = null;
  }

  execute() {
    if (existsSync(this.filePath)) {
      this._backup = readFileSync(this.filePath, 'utf-8');
      unlinkSync(this.filePath);
    }
  }

  undo() {
    if (this._backup !== null) {
      writeFileSync(this.filePath, this._backup, 'utf-8');
    }
  }
}

export class CompositeCommand {
  constructor() {
    this.commands = [];
    this.description = 'コンポジットコマンド';
  }

  addCommand(command) {
    this.commands.push(command);
  }

  execute() {
    for (const command of this.commands) {
      command.execute();
    }
  }

  undo() {
    // 逆順に undo する
    for (const command of [...this.commands].reverse()) {
      command.undo();
    }
  }
}

export class SlickButton {
  constructor(callback) {
    this.callback = callback;
  }

  click() {
    this.callback();
  }
}
